src/db_options.rs

```diff
@@ -3872,26 +3888,43 @@ impl ReadOptions {
     /// only the most recent version visible to timestamp is returned.
     /// The user-specified timestamp feature is still under active development,
     /// and the API is subject to change.
-    pub fn set_timestamp<S: AsRef<[u8]>>(&mut self, ts: S) {
-        let ts = ts.as_ref();
+    pub fn set_timestamp<S: Into<Vec<u8>>>(&mut self, ts: S) {
+        self.set_timestamp_impl(Some(ts.into()))
+    }
+
+    fn set_timestamp_impl(&mut self, ts: Option<Vec<u8>>) {
+        let (ptr, len) = if let Some(ref ts) = ts {
+            (ts.as_ptr() as *const c_char, ts.len())
+        } else if self.timestamp.is_some() {
+            // The stored timestamp is a `Some` but we're updating it to a `None`.
+            // This means to cancel a previously set timestamp.
+            // To do this, use a null pointer and zero length.
+            (std::ptr::null(), 0)
+        } else {
+            return;
+        };
+        self.timestamp = ts;
         unsafe {
-            ffi::rocksdb_readoptions_set_timestamp(
-                self.inner,
-                ts.as_ptr() as *const c_char,
-                ts.len() as size_t,
-            );
+            ffi::rocksdb_readoptions_set_timestamp(self.inner, ptr, len);
         }
```

Comment:
We use exactly the same approach as in `ReadOptions::set_iterator_{lower,upper}_bound`